cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,258 @@
+	S3Endpoint     string
+	S3AccessKeyID  string
+	S3AccessKey    string
+	S3BucketName   string
+	S3Region       string
+	RecoverTime    string
+	RecoverType    string
+	BackupName     string
+	GTIDSet        string
+}
+
+type Storage interface {
+	GetObject(objectName string) (io.Reader, error)
+	PutObject(name string, data io.Reader) error
+	ListObjects(prefix string) []string
+}
+
+type RecoverType string
+
+func New(c Config) (*Recoverer, error) {
+	s3, err := storage.NewS3(c.S3Endpoint, c.S3AccessKeyID, c.S3AccessKey, c.S3BucketName, c.S3Region, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "new storage manager")
+	}
+	return &Recoverer{
+		storage:        s3,
+		recoverTime:    c.RecoverTime,
+		pxcUser:        c.PXCUser,
+		pxcServiceName: c.PXCServiceName,
+		recoverType:    RecoverType(c.RecoverType),
+		backupName:     c.BackupName,
+	}, nil
+}
+
+const (
+	Latest      RecoverType = "latest"      // recover to the latest existing binlog
+	Date        RecoverType = "date"        // recover to exact date
+	Transaction RecoverType = "transaction" // recover to needed trunsaction
+	Skip        RecoverType = "skip"        // skip transactions
+)
+
+func (r *Recoverer) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+r.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost + "." + r.pxcServiceName, nil
+}
+
+func (r *Recoverer) Run() error {
+	startGTID, err := r.getLastBackupGTID()
+	if err != nil {
+		return errors.Wrap(err, "get last gtid")
+	}
+	err = r.setBinlogs(startGTID)
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+	err = r.recover()
+	if err != nil {
+		return errors.Wrap(err, "recover")
+	}
+
+	return nil
+}
+
+func (r *Recoverer) recover() error {
+	host, err := r.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+
+	flags := ""
+	endTime := time.Time{}
+	// TODO: add logic for all types
+	switch r.recoverType {
+	case Skip:
+		flags = " --exclude-gtids=" + r.gtidSet
+	case Transaction:
+	case Date:
+		flags = ` --stop-datetime="` + r.recoverTime + `"`
+
+		format := "2006-01-02 15:04:05"
+		t, err := time.Parse(format, r.recoverTime)
+		if err != nil {
+			return errors.Wrap(err, "parse date")
+		}
+		endTime = t
+	case Latest:
+	default:
+		return errors.New("wrong recover type")
+	}
+
+	for _, binlog := range r.binlogs {
+		log.Println("working with", binlog)
+
+		if r.recoverType == Date {
+			binlogArr := strings.Split(binlog, ":")
```

Comment:
`:` is prohibited symbol on some filesystems
```suggestion
			binlogArr := strings.Split(binlog, "_")
```